common/tools/eslint-plugin-azure-sdk/src/rules/ts-doc-internal-private-member.ts

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * @file Rule to check if a private class member is tagged with @internal
+ * @author Hamsa Shankar
+ */
+
+import { ParserServices, TSESTree } from "@typescript-eslint/experimental-utils";
+import { ParserWeakMapESTreeToTSNode } from "@typescript-eslint/typescript-estree/dist/parser-options";
+import { Rule } from "eslint";
+import { Node } from "estree";
+import { readFileSync } from "fs";
+import { sync as globSync } from "glob";
+import { relative } from "path";
+import { Modifier, SyntaxKind } from "typescript";
+import { getRuleMetaData } from "../utils";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+/**
+ * Helper method for reporting on a node
+ * @param node the Node being operated on
+ * @param context the ESLint runtime context
+ * @param converter a converter from TSESTree Nodes to TSNodes
+ * @param typeChecker the TypeScript TypeChecker
+ * @throws if the Node passes throught the initial checks and has an internal tag
+ */
+const reportInternal = (
+  node: Node,
+  context: Rule.RuleContext,
+  converter: ParserWeakMapESTreeToTSNode,
+): void => {
+  const tsNode = converter.get(node as TSESTree.Node) as any;
+
+  const modifiers = converter.get(node as TSESTree.Node).modifiers;
+  // if type is internal and has a TSDoc and is a private member
+  if (tsNode.jsDoc !== undefined
+    && modifiers !== undefined && modifiers.some(
+      (modifier: Modifier): boolean => modifier.kind === SyntaxKind.PrivateKeyword)
+  ) {
+    // fetch all tags
+    let TSDocTags: string[] = [];
+    tsNode.jsDoc.forEach((TSDocComment: any): void => {
+      TSDocTags = TSDocTags.concat(
+        TSDocComment.tags !== undefined
+          ? TSDocComment.tags.map((TSDocTag: any): string => TSDocTag.tagName.escapedText)
```

Comment:
A performance optimization could be to use reduce (via a separate function, to avoid adding more complexity here) to retrieve only the `(internal)` tags directly during this loop. Then the conditional on line 55 could be limited to check if the array is empty or not.

Reply:
I added a filtering function here since it was a single line and then updated the conditional to check the length but let me know if I should change it back or do something else

Reply:
Do we need to create the `TSDocTags` array? Couldn't we loop through tags and report when there's a match?

Reply:
Thanks for the suggestion, I just updated it to loop through and report whenever it sees an internal tag. I think that's better than what I had before